File: java/src/main/java/com/dsa/algorithms/DFS.java
package com.dsa.algorithms;

import java.util.*;
import java.util.function.Consumer;

/**
 * Depth-First Search implementation.
 * Time: O(V + E)
 * Space: O(V)
 */
public class DFS {

    public static <T> List<T> traverse(Graph<T> graph, T start) {
        return traverse(graph, start, null);
    }

    public static <T> List<T> traverse(Graph<T> graph, T start, Consumer<T> callback) {
        List<T> result = new ArrayList<>();

        if (!graph.hasVertex(start)) {
            return result;
        }

        Set<T> visited = new HashSet<>();
        Deque<T> stack = new ArrayDeque<>();

        stack.push(start);

        while (!stack.isEmpty()) {
            T current = stack.pop();

            if (visited.contains(current)) {
                continue;
            }

            visited.add(current);
            result.add(current);

            if (callback != null) {
                callback.accept(current);
            }

            List<T> neighbors = graph.neighbors(current);
            for (int i = neighbors.size() - 1; i >= 0; i--) {
                T neighbor = neighbors.get(i);
                if (!visited.contains(neighbor)) {
                    stack.push(neighbor);
                }
            }
        }

        return result;
    }

    public static <T> List<T> traverseRecursive(Graph<T> graph, T start) {
        return traverseRecursive(graph, start, null);
    }

    public static <T> List<T> traverseRecursive(Graph<T> graph, T start, Consumer<T> callback) {
        List<T> result = new ArrayList<>();

        if (!graph.hasVertex(start)) {
            return result;
        }

        Set<T> visited = new HashSet<>();
        dfsRecursive(graph, start, visited, result, callback);
        return result;
    }

    private static <T> void dfsRecursive(
            Graph<T> graph, T current, Set<T> visited, List<T> result, Consumer<T> callback) {
        visited.add(current);
        result.add(current);

        if (callback != null) {
            callback.accept(current);
        }

        for (T neighbor : graph.neighbors(current)) {
            if (!visited.contains(neighbor)) {
                dfsRecursive(graph, neighbor, visited, result, callback);
            }
        }
    }

    public static <T> Optional<List<T>> findPath(Graph<T> graph, T start, T end) {
        if (!graph.hasVertex(start) || !graph.hasVertex(end)) {
            return Optional.empty();
        }

        if (start.equals(end)) {
            return Optional.of(Collections.singletonList(start));
        }

        Set<T> visited = new HashSet<>();
        List<T> path = new ArrayList<>();

        if (dfsPath(graph, start, end, visited, path)) {
            return Optional.of(path);
        }
        return Optional.empty();
    }

    private static <T> boolean dfsPath(
            Graph<T> graph, T current, T end, Set<T> visited, List<T> path) {
        visited.add(current);
        path.add(current);

        if (current.equals(end)) {
            return true;
        }

        for (T neighbor : graph.neighbors(current)) {
            if (!visited.contains(neighbor)) {
                if (dfsPath(graph, neighbor, end, visited, path)) {
                    return true;
                }
            }
        }

        path.remove(path.size() - 1);
        return false;
    }

    public static <T> boolean hasCycle(Graph<T> graph, T start) {
        if (!graph.hasVertex(start)) {
            return false;
        }

        Set<T> visited = new HashSet<>();
        Set<T> recStack = new HashSet<>();

        return detectCycle(graph, start, visited, recStack);
    }

    private static <T> boolean detectCycle(
            Graph<T> graph, T current, Set<T> visited, Set<T> recStack) {
        visited.add(current);
        recStack.add(current);

        for (T neighbor : graph.neighbors(current)) {
            if (!visited.contains(neighbor)) {
                if (detectCycle(graph, neighbor, visited, recStack)) {
                    return true;
                }
            } else if (recStack.contains(neighbor)) {
                return true;
            }
        }

        recStack.remove(current);
        return false;
    }
}

File: zig/src/algorithms/dfs.zig
const std = @import("std");
const Allocator = std.mem.Allocator;
const Graph = @import("graph.zig").Graph;

/// Depth-first search algorithms.
/// Time: O(V + E)
/// Space: O(V)

pub fn traverse(comptime T: type, graph: *const Graph(T), start: T, allocator: Allocator) !std.ArrayList(T) {
    var result = std.ArrayList(T).init(allocator);
    errdefer result.deinit();

    if (!graph.hasVertex(start)) return result;

    var visited = std.AutoHashMap(T, void).init(allocator);
    defer visited.deinit();

    var stack = std.ArrayList(T).init(allocator);
    defer stack.deinit();

    try stack.append(start);

    while (stack.items.len > 0) {
        const vertex = stack.pop();

        if (visited.contains(vertex)) continue;

        try visited.put(vertex, {});
        try result.append(vertex);

        if (graph.neighbors(vertex)) |neighbors| {
            var it = neighbors.keyIterator();
            while (it.next()) |neighbor| {
                if (!visited.contains(neighbor.*)) {
                    try stack.append(neighbor.*);
                }
            }
        }
    }

    return result;
}

pub fn findPath(comptime T: type, graph: *const Graph(T), start: T, end: T, allocator: Allocator) !?std.ArrayList(T) {
    if (!graph.hasVertex(start) or !graph.hasVertex(end)) return null;
    if (start == end) {
        var result = std.ArrayList(T).init(allocator);
        try result.append(start);
        return result;
    }

    var visited = std.AutoHashMap(T, void).init(allocator);
    defer visited.deinit();

    var parent = std.AutoHashMap(T, T).init(allocator);
    defer parent.deinit();

    var stack = std.ArrayList(T).init(allocator);
    defer stack.deinit();

    try stack.append(start);

    while (stack.items.len > 0) {
        const vertex = stack.pop();

        if (visited.contains(vertex)) continue;
        try visited.put(vertex, {});

        if (vertex == end) {
            var path = std.ArrayList(T).init(allocator);
            var current = end;

            while (current != start) {
                try path.append(current);
                current = parent.get(current).?;
            }
            try path.append(start);

            std.mem.reverse(T, path.items);
            return path;
        }

        if (graph.neighbors(vertex)) |neighbors| {
            var it = neighbors.keyIterator();
            while (it.next()) |neighbor| {
                if (!visited.contains(neighbor.*)) {
                    try parent.put(neighbor.*, vertex);
                    try stack.append(neighbor.*);
                }
            }
        }
    }

    return null;
}

pub fn hasCycle(comptime T: type, graph: *const Graph(T), allocator: Allocator) !bool {
    var visited = std.AutoHashMap(T, void).init(allocator);
    defer visited.deinit();

    var rec_stack = std.AutoHashMap(T, void).init(allocator);
    defer rec_stack.deinit();

    var it = graph.adjacency_list.keyIterator();
    while (it.next()) |vertex| {
        if (!visited.contains(vertex.*)) {
            if (try hasCycleUtil(T, graph, vertex.*, &visited, &rec_stack)) {
                return true;
            }
        }
    }

    return false;
}

fn hasCycleUtil(comptime T: type, graph: *const Graph(T), vertex: T, visited: *std.AutoHashMap(T, void), rec_stack: *std.AutoHashMap(T, void)) !bool {
    try visited.put(vertex, {});
    try rec_stack.put(vertex, {});

    if (graph.neighbors(vertex)) |neighbors| {
        var it = neighbors.keyIterator();
        while (it.next()) |neighbor| {
            if (!visited.contains(neighbor.*)) {
                if (try hasCycleUtil(T, graph, neighbor.*, visited, rec_stack)) {
                    return true;
                }
            } else if (rec_stack.contains(neighbor.*)) {
                return true;
            }
        }
    }

    _ = rec_stack.remove(vertex);
    return false;
}

test "DFS traverse" {
    const allocator = std.testing.allocator;
    var graph = Graph(i32).init(allocator, false);
    defer graph.deinit();

    try graph.addEdge(1, 2);
    try graph.addEdge(1, 3);
    try graph.addEdge(2, 4);

    var result = try traverse(i32, &graph, 1, allocator);
    defer result.deinit();

    try std.testing.expectEqual(@as(usize, 4), result.items.len);
    try std.testing.expectEqual(@as(i32, 1), result.items[0]);
}
